src/workerd/util/sqlite.c++

```diff
@@ -438,16 +447,62 @@ kj::Own<sqlite3_stmt> SqliteDatabase::prepareSql(
           } else {
             SQLITE_CALL_FAILED("sqlite3_step()", err);
           }
-
-          // Reduce `sqlCode` to include only what we haven't already executed.
-          sqlCode = kj::StringPtr(tail, sqlCode.end());
-          continue;
         }
-        break;
+
+        // Now set statementResult as the "result" to be either returned or step()-ed through
+        result = statementResult;
+      }
+
+      // Reduce `sqlCode` to include only what we haven't already executed.
+      sqlCode = kj::StringPtr(tail, sqlCode.end());
+    }
+
+    // After processing the entire input, throw if we never got a single result
+    SQLITE_REQUIRE(result != nullptr, "SQL code did not contain a complete statement.", sqlCode);
+    return ownSqlite(result);
+  }
+}
+
+int SqliteDatabase::ingestSql(Regulator& regulator, kj::StringPtr sqlCode) {
+  // Set up the regulator that will be used for authorizer callbacks while preparing this
+  // statement.
+  KJ_ASSERT(currentRegulator == nullptr, "recursive prepareSql()?");
+  KJ_DEFER(currentRegulator = nullptr);
+  currentRegulator = regulator;
+  const char* start = sqlCode.begin();
+  const char* tail = start;
+
+  SQLITE_REQUIRE(sqlite3_get_autocommit(db) == 0, "sql.ingest must be run inside a transaction");
+
+  // While there are still valid
+  while (tail != sqlCode.end() && sqlite3_complete_inner(tail, 1)) {
+    // Compile the next statement in the SQL
+    sqlite3_stmt* result = nullptr;
+    SQLITE_CALL(sqlite3_prepare_v3(db, sqlCode.begin(), sqlCode.size(), 0, &result, &tail));
```

Comment:
There's a bunch of code here duplicating stuff in `prepareSql()`. Could sqlite3_complete_inner() instead return the split point so we can split off exactly one statement and pass it to prepareSql()? Or could we add a new `Multi` enum for `prepareSql()` which instructs it to parse a single statement and then not fail if there's data left over, but instead maybe modify `sqlCode` in-place to remove the prefix?

Reply:
Have refactored, `ingestSql` now calls `prepareSql` and it's a bit nicer. It could be better with a bit more refactoring but I'll clean that up once I'm confident this version works re: string encoding.